### PR TITLE
commit visualization in commit details page

### PR DESCRIPTION
**Pull request by karthikjeeyar**
## Fixes 

https://issues.redhat.com/browse/HACBS-826

## Description

Adds commit visualization in commit details page.

## Type of change

- [x] Feature

## Screen shots / Gifs for design review 

image

image

## How to test or reproduce?

Yamls to create test setup:

Test and Environments are here - https://gist.github.com/karthikjeeyar/ce5a1038209565e5ecf0f3a9719609be
Release Setup yamls -  https://github.com/karthikjeeyar/hacbs-demos/tree/main/m7/release

### CODE REVIEW

**Comment by karthikjeeyar** [comment]
cc: @christianvogt @jeff-phillips-18

**Comment by karthikjeeyar** [comment]
cc: @Ranelim @MariaLeonova

**Comment by MariaLeonova** [comment]
@Ranelim @andrew-ronaldson  let's update the icons as part of this PR?
@karthikjeeyar what format would you want them in? I think we're good with the github icons, while the other two (pipeline, environment) need to be updated.

**Comment by karthikjeeyar** [comment]
@MariaLeonova If it is not a patternfly icons, you can provide me in SVG format.

**Comment by MariaLeonova** [comment]
@karthikjeeyar this one is for pipelines 
pipelineIcon

**Comment by jeff-phillips-18** [comment]
@karthikjeeyar I have added the pipeline SVG as part of #248

**Comment by christianvogt** [comment]
When testing this change the sizing of other visualizations seems broken. At the same time I cannot reproduce this on console.dev:

image
image

**Comment by karthikjeeyar** [comment]
Rebase, Updated Icons and fixed the size issues in other visualisations.

File: src/hacbs/components/topology/factories/VisualizationFactory.tsx [hunk 1]
```diff
@@ -80,18 +84,26 @@ const VisualizationFactory: React.FC<VisualizationFactoryProps> = ({
       controller.fromModel(model, model.graph.layout === layoutRef.current);
       layoutRef.current = model.graph.layout;
       controller.getGraph().layout();
+      if (fitToScreen) {
+        controller.getGraph().fit(90);
+      }
```
Comment (christianvogt) [inline src/hacbs/components/topology/factories/VisualizationFactory.tsx:89]: I don't see `fitToScreen` being used by any caller of this function.

Furthermore if `fitToScreen` changes, this effect does a whole lot more than what's needed to fit to screen. 

If you still need this prop, move this to a separate effect.

File: src/hacbs/components/Commits/tabs/CommitDetails/CommitComponentFactory.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  TaskEdge,
+  withPanZoom,
+} from '@patternfly/react-topology';
+import { NodeType } from '../../../ApplicationDetails/tabs/overview/visualization/const';
+import CommitWorkflowNode from '../../../ApplicationDetails/tabs/overview/visualization/nodes/CommitWorkflowNode';
+
+export const commitComponentFactory: ComponentFactory = (kind: ModelKind, type: string) => {
+  switch (kind) {
+    case ModelKind.graph:
+      return withPanZoom()(GraphComponent);
```
Comment (christianvogt) [inline src/hacbs/components/Commits/tabs/CommitDetails/CommitComponentFactory.tsx:15]: pan zoom is fine for now until UX gives a final decision if this needs to be removed in lieu of scroll bars.

File: src/hacbs/components/ApplicationDetails/tabs/overview/visualization/utils/visualization-utils.ts [hunk 3]
```diff
@@ -44,8 +119,93 @@ export const getLatestResource = (resources = []) =>
...
+        width: getLabelWidth(label),
+      };
+
+      workflowDag.addEdges(id, wData, runBefore, runAfter);
+    } else {
+      const rootResources = !runAfterResourceKey
+        ? resources
+        : (resources as []).filter((r) => {
+            return !get(r, runAfterResourceKey);
+          });
```
Comment (christianvogt) [inline src/hacbs/components/ApplicationDetails/tabs/overview/visualization/utils/visualization-utils.ts:152]: `r` is of type `never` here.
Avoid casting `[]` by doing proper checks.

File: src/hacbs/components/ApplicationDetails/tabs/overview/visualization/hooks/useCommitWorkflowData.ts [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+  if (!allResourcesLoaded) {
+    return [[], allResourcesLoaded];
+  }
+
+  const commitComponents = buildPipelines.map(
+    (bp) => bp.metadata.labels[PipelineRunLabel.COMMIT_COMPONENT_LABEL],
+  );
+
+  const commitWorkflowData: { [key: string]: any } = components.reduce(
```
Comment (christianvogt) [inline src/hacbs/components/ApplicationDetails/tabs/overview/visualization/hooks/useCommitWorkflowData.ts:76]: There's opportunity to memoize some values here such that we don't need to recompute everything if only a single value changes.

File: src/hacbs/components/ApplicationDetails/tabs/overview/visualization/utils/visualization-utils.ts (hunk 3 above)
Comment (karthikjeeyar) [reply src/hacbs/components/ApplicationDetails/tabs/overview/visualization/utils/visualization-utils.ts:152]: Typescript throws an error when we use Union of array types  for methods like filter(), find(), every() and there is an [open issue](https://github.com/microsoft/TypeScript/issues/44373) on Typescript to provide a fix for this. :(

File: src/hacbs/components/topology/factories/VisualizationFactory.tsx (hunk 1 above)
Comment (karthikjeeyar) [reply src/hacbs/components/topology/factories/VisualizationFactory.tsx]: I removed it as it is not used in other graphs

File: src/hacbs/components/ApplicationDetails/tabs/overview/visualization/hooks/useCommitWorkflowData.ts (hunk 4 above)
Comment (christianvogt) [reply src/hacbs/components/ApplicationDetails/tabs/overview/visualization/hooks/useCommitWorkflowData.ts:76]: That being said, it's something that can be improved on in future PRs and not needed right away.

File: src/hacbs/components/ApplicationDetails/tabs/overview/visualization/utils/visualization-utils.ts (hunk 3 above)
Comment (christianvogt) [reply src/hacbs/components/ApplicationDetails/tabs/overview/visualization/utils/visualization-utils.ts]: Ah that's too bad :(
Ok work around is fine for now. Thanks!

File: src/hacbs/components/Commits/tabs/CommitDetails/CommitComponentFactory.tsx (hunk 2 above)
Comment (jeff-phillips-18) [reply src/hacbs/components/Commits/tabs/CommitDetails/CommitComponentFactory.tsx:15]: Inconsistent with the application overview graph. Whichever way is finally decided, these should be made consistent.

File: src/hacbs/components/Commits/tabs/CommitDetails/CommitVisualization.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  defaultControlButtonsOptions,
+  ModelKind,
+  TopologyControlBar,
+} from '@patternfly/react-topology';
+import { useCommitWorkflowData } from '../../../ApplicationDetails/tabs/overview/visualization/hooks/useCommitWorkflowData';
+import { getTopologyNodesEdges } from '../../../ApplicationDetails/tabs/overview/visualization/utils/visualization-utils';
+import { layoutFactory, PipelineLayout, VisualizationFactory } from '../../../topology/factories';
+//
+import './CommitVisualization.scss';
+import { commitComponentFactory } from './CommitComponentFactory';
```
Comment (jeff-phillips-18) [inline src/hacbs/components/Commits/tabs/CommitDetails/CommitVisualization.tsx:14]: Wonder why the linter didn't pickup the import order here

File: src/hacbs/components/Commits/tabs/CommitDetails/CommitVisualization.tsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    return null;
+  }
+  return (
+    <div className="hacbs-commit-graph" data-testid="hacbs-commit-graph">
+      <VisualizationFactory
+        componentFactory={commitComponentFactory}
+        layoutFactory={layoutFactory}
+        model={model}
+        fullHeight
+        controlBar={(controller) => (
```
[the remark this answers is not in the document]
Comment (karthikjeeyar) [reply src/hacbs/components/Commits/tabs/CommitDetails/CommitVisualization.tsx:40]: Yeah, waiting on UX for final update on this and pan zoom feature as well.

File: src/hacbs/components/Commits/tabs/CommitDetails/CommitVisualization.tsx (hunk 5 above)
Comment (karthikjeeyar) [reply src/hacbs/components/Commits/tabs/CommitDetails/CommitVisualization.tsx:12]: Fixed it.

**Review by jeff-phillips-18 — APPROVED** [review]

**Pull request merged**